### ISSUE TITLE
umount/mount modifies readonly filesystem

### ISSUE THREAD

**Issue by stuartthebruce** [zfsonlinux/zfs#6741]
### System information
Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | Scientific Linux
Distribution Version    | 7.4
Linux Kernel                 | 3.10.0-693.2.2.el7.x86_64
Architecture                 | x86_64
ZFS Version                  | 0.7.2-1
SPL Version                  | 0.7.2-1

### Describe the problem you're observing
unmounting and mounting a reports a readonly filesystem has been modified

### Describe how to reproduce the problem
```
[root@dcc3 ~]# zfs get readonly,written dcc/data
NAME      PROPERTY  VALUE    SOURCE
dcc/data  readonly  on       local
dcc/data  written   0        -
[root@dcc3 ~]# zfs umount dcc/data && zfs mount dcc/data
[root@dcc3 ~]# zfs get readonly,written dcc/data
NAME      PROPERTY  VALUE    SOURCE
dcc/data  readonly  on       local
dcc/data  written   141M     -
```

### Include any warning/errors/backtraces from the system logs
This results in zfs receiving failing to accept an incremental snapshot, "cannot receive incremental stream: destination XXX has been modified"

**Comment by stuartthebruce** [comment]
This issue persists even after setting atime=off

**Comment by ronnyegner** [comment]
> This results in zfs receiving failing to accept an incremental snapshot, "cannot receive incremental stream: destination XXX has been modified"

I think you can work around this by adding "-F" to the recv.

**Comment by siebenmann** [comment]
It's possible that this is due to pending deletes being processed when the filesystem is mounted. You get pending deletes if you open a file and keep it open in a process, then delete the file on disk; the file isn't accessible at the filesystem level, but it can't be removed until the process closes it. ZFS implements this with a per-filesystem queue of pending deletes, and this queue is captured in snapshots and transferred through 'zfs send | zfs recv'. If you mount a filesystem with a pending delete queue, ZFS has traditionally processed the queue even if the filesystem was mounted read-only, creating a divergence between the mounted filesystem and the snapshot.

It's possible to use zdb to see if you have such a pending delete queue; a discussion of how to do this is part of [my blog entry about ZFS delete queues](https://utcc.utoronto.ca/~cks/space/blog/solaris/ZFSDeleteQueue) (and [see also](https://utcc.utoronto.ca/~cks/space/blog/solaris/ZFSDeleteQueueSnapshots) for a discussion of how this interacts with snapshots and so on).

**Comment by dward** [comment]
After some troubleshooting, we have finally determined the series of events that put us in this same situation.

TLDR; The z_upgrade process runs on a readonly dataset to add userobj/project tracking to the dataset, which modified the dataset.

1. The replica's dataset or last full replication had to be creating prior to the existence of the userobj_accounting feature.
2. The userobj_accounting feature gets enabled on the origin pool and replica pool.
3. Feature incremental replications will work until the replica dataset is unmounted and remounted. The mount process appears to spawn off the z_upgrade process which will modify the dataset readonly dataset.

A full replication will fix the issue as a newly created dataset on the replica will have the OBJSET_FLAG_USEROBJACCOUNTING_COMPLETE set at creation time.

But, if you're dealing with terabytes or petabytes of data, a full replication is quite cumbersome. Should the z_upgrade process even run on readonly datasets?

**Comment by dward** [comment]
To correct something above. 

It appears that the code *does* in fact attempt to skip readonly datasets.  The readonly detection method only appears to work when /sbin/mount.zfs utility installed and for some reason we didn't have it in our installation.

**Comment by vienfla** [comment]
Hi,
any updates on this issue?
Seems the problem is still happening in 2021. 

Rebooting the machine breaks the synchronisation:
cannot receive incremental stream ... destination has been modified
we do have the utility /sbin/mount.zfs

Thanks

**Comment by stuartthebruce** [comment]
I think this is still worth fixing.

**Comment by ghjs** [comment]
I would like to report this issue is still present on ZFS tag 2.2.0-rc4.

The details of dward's https://github.com/openzfs/zfs/issues/6741#issuecomment-707815791 match what I'm seeing.

Finally finding time to try to make sense of what I've been seeing with a nearly 10 year old, large ZFS backup zpool that calls out to remote systems to pull in snapshots, this issue was the best match.  Plenty of the remote zpools, besides the backup zpool, were in service prior to userobj_accounting/project_quota being a zpool feature.

Some of the remote zpools datasets that this zpool is an off-site backup for have both large content (10s/100s TiBs) and large number of files (many millions) and doing a re-pull of a full replica to avoid this per-zpool import scan seems wasteful, so for now I've set the canmount property to off so the import doesn't spawn the z_upgrade thread (dmu_objset_id_quota_upgrade_cb in the kernel call stack) to crawl the roughly ~20 impacted ZFS filesystem datasets.

While this process firing and running on datasets with readonly properties is a bug, I also feel it would be appropriate for the userobj_accounting/project_quota updates from the origin zpool are also addressed during an incremental snapshot update to the receiving zpool, if the receiving zpool has the feature active, and stays.  This is so future imports will have the `os_flags |= OBJSET_FLAG_<TYPE>_COMPLETE` set to avoid the scan and have that data available on the replica zpool to query with `zfs userspace` and not have to waste IOPs/bandwidth scanning for it when it's later lost by the next received incremental snapshot update from the origin zpool.  It's not clear to my why a full replica snapshot send with the feature active on the dataset can address this issue with ongoing incrementals from there, while zpool replica setups for backup purposes predating the feature (but with the feature active on both zpools) can't sync up based on the origin's data structures to track these features.
